File: solutions/python/src/scheduler/graph.py
from collections import defaultdict, deque
from .models import Status, Task

class DependencyGraph:
    def __init__(self, tasks: list[Task]) -> None:
        self.tasks={t.id:t for t in tasks}
        for t in tasks:
            missing=t.dependencies-self.tasks.keys()
            if missing: raise KeyError(f"unknown dependencies for {t.id}: {sorted(missing)}")
        self.topological_order()
    def topological_order(self) -> list[str]:
        indegree={tid:0 for tid in self.tasks}; children:dict[str,list[str]]=defaultdict(list)
        for t in self.tasks.values():
            indegree[t.id]=len(t.dependencies)
            for dep in t.dependencies: children[dep].append(t.id)
        queue=deque(sorted(k for k,v in indegree.items() if v==0)); order=[]
        while queue:
            node=queue.popleft(); order.append(node)
            for child in sorted(children[node]):
                indegree[child]-=1
                if indegree[child]==0: queue.append(child)
        if len(order)!=len(self.tasks): raise ValueError("dependency cycle")
        return order
    def ready_ids(self) -> list[str]:
        return [t.id for t in self.tasks.values() if t.status in {Status.PENDING,Status.READY} and all(self.tasks[d].status==Status.SUCCEEDED for d in t.dependencies)]

File: solutions/cpp/src/graph.cpp
#include "scheduler/graph.hpp"
#include <algorithm>
#include <deque>
#include <unordered_map>
namespace scheduler {
std::vector<std::string> DependencyGraph::topological_order()const{auto tasks=catalog_.list();std::unordered_map<std::string,int> degree;std::unordered_map<std::string,std::vector<std::string>> children;for(auto r:tasks)degree[r.get().id]=0;for(auto r:tasks){for(auto const&d:r.get().dependencies){if(!degree.contains(d))throw std::invalid_argument("unknown dependency");++degree[r.get().id];children[d].push_back(r.get().id);}}std::vector<std::string> zeros;for(auto const&[id,n]:degree)if(n==0)zeros.push_back(id);std::ranges::sort(zeros);std::deque<std::string> q(zeros.begin(),zeros.end());std::vector<std::string> out;while(!q.empty()){auto id=q.front();q.pop_front();out.push_back(id);auto& cs=children[id];std::ranges::sort(cs);for(auto const&c:cs)if(--degree[c]==0)q.push_back(c);}if(out.size()!=tasks.size())throw std::logic_error("cycle");return out;}
std::vector<std::string> DependencyGraph::ready_ids()const{std::vector<std::string> out;for(auto r:catalog_.list()){auto const&t=r.get();if(t.status!=Status::pending&&t.status!=Status::ready)continue;bool ready=std::ranges::all_of(t.dependencies,[&](auto const&d){return catalog_.get(d).status==Status::succeeded;});if(ready)out.push_back(t.id);}return out;}
}
